[Bug] Esplora reqwest blockchain tests are failing

### Bug

It seems that `esplora-reqwest` blockchain tests have been failing for some time now.

The CI missed it because all the esplora tests were being skipped. 

The major sources of failure are `async` related. Especially calling `async` functions in `bdk_blockchain_tests` macro. 

Error sample 
```
error[E0599]: no method named `unwrap` found for opaque type `impl futures::Future` in the current scope
   --> src/testutils/blockchain_tests.rs:875:52
    |
875 |                   wallet.sync(noop_progress(), None).unwrap();
    |                                                      ^^^^^^ method not found in `impl futures::Future`
    | 
   ::: src/blockchain/esplora/mod.rs:125:1
    |
125 | / crate::bdk_blockchain_tests! {
126 | |     fn test_instance(test_client: &TestClient) -> EsploraBlockchain {
127 | |         EsploraBlockchain::new(&format!("http://{}",test_client.electrsd.esplora_url.as_ref().unwrap()), 20)
128 | |     }
129 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider `await`ing on the `Future` and calling the method on its `Output`
    |
875 |                 wallet.sync(noop_progress(), None).await.unwrap();
    |                                                    ^^^^^^

error: aborting due to 58 previous errors

Some errors have detailed explanations: E0195, E0308, E0599.
For more information about an error, try `rustc --explain E0195`.
```

### Reproduce: 

Edit test-esplora feature in cargo like `test-esplora = ["use-esplora-reqwest", "electrsd/legacy", "electrsd/esplora_a33e97e1", "test-blockchains"]`. 

then run `cargo test --features test-esplora`

### Analysis
As far as I understand, its happening because in the test macro we are calling wallet APIs in non async mode, whereas, for `esplora-reqwest` these are async functions.

https://github.com/bitcoindevkit/bdk/blob/721748e98f95f8f97d508322ae7871ec4c49f0c4/src/testutils/blockchain_tests.rs#L378-L380

### Thoughts
I tried putting in the `maybe_async` and `maybe_await` macros where applicable. But `#[test]` are not allowed to be async (or at least the vanilla flavour of them).

we can have async tests but with using `tokio` directly. But then, they cannot be non-async anymore.

Or maybe we can create some proc macros like `#[maybe_async_test]`? 

I noticed this issue few weeks back in bdk-cli https://github.com/bitcoindevkit/bdk-cli/pull/36#issue-702356592. Was confused why I couldn't reproduce it in BDK. The reason is the current `test-esplora` feature is broken, and is not compiling the blockchian tests. This is now fixed with https://github.com/bitcoindevkit/bdk/pull/430

--- a/.github/workflows/cont_integration.yml
+++ b/.github/workflows/cont_integration.yml
@@ -78,15 +78,20 @@ jobs:
         run: cargo test --features test-md-docs --no-default-features -- doctest::ReadmeDoctests
 
   test-blockchains:
-    name: Test ${{ matrix.blockchain.name }}
+    name: Blockchain ${{ matrix.blockchain.features }}
     runs-on: ubuntu-20.04
     strategy:
       fail-fast: false
       matrix:
         blockchain:
           - name: electrum
+            features: test-electrum
           - name: rpc
+            features: test-rpc
           - name: esplora
+            features: test-esplora,use-esplora-reqwest
+          - name: esplora
+            features: test-esplora,use-esplora-ureq    
     steps:
       - name: Checkout
         uses: actions/checkout@v2
@@ -104,8 +109,8 @@ jobs:
           toolchain: stable
           override: true
       - name: Test
-        run: cargo test --features test-${{ matrix.blockchain.name }} ${{ matrix.blockchain.name }}::bdk_blockchain_tests
-
+        run: cargo test --no-default-features --features ${{ matrix.blockchain.features }} ${{ matrix.blockchain.name }}::bdk_blockchain_tests
+  
   check-wasm:
     name: Check WASM
     runs-on: ubuntu-16.04
@@ -138,7 +143,6 @@ jobs:
         run: rustup update
       - name: Check
         run: cargo check --target wasm32-unknown-unknown --features use-esplora-reqwest --no-default-features
-
 
   fmt:
     name: Rust fmt

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -12,7 +12,7 @@ readme = "README.md"
 license = "MIT OR Apache-2.0"
 
 [dependencies]
-bdk-macros = "0.5"
+bdk-macros = { path = "macros"} # TODO: Change this to version number after next release.
 log = "^0.4"
 miniscript = "^6.0"
 bitcoin = { version = "^0.27", features = ["use-serde", "base64"] }
@@ -37,6 +37,10 @@ bitcoinconsensus = { version = "0.19.0-3", optional = true }
 
 # Needed by bdk_blockchain_tests macro
 core-rpc = { version = "0.14", optional = true }
+
+# Platform-specific dependencies
+[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
+tokio = { version = "1", features = ["rt"] }
 
 [target.'cfg(target_arch = "wasm32")'.dependencies]
 async-trait = "0.1"
@@ -70,7 +74,7 @@ rpc = ["core-rpc"]
 async-interface = ["async-trait"]
 electrum = ["electrum-client"]
 # MUST ALSO USE `--no-default-features`.
-use-esplora-reqwest = ["async-interface", "esplora", "reqwest", "futures"]
+use-esplora-reqwest = ["esplora", "reqwest", "futures"]
 use-esplora-ureq = ["esplora", "ureq"]
 # Typical configurations will not need to use `esplora` feature directly.
 esplora = []
@@ -80,7 +84,7 @@ esplora = []
 test-blockchains = ["core-rpc", "electrum-client"]
 test-electrum = ["electrum", "electrsd/electrs_0_8_10", "test-blockchains"]
 test-rpc = ["rpc", "electrsd/electrs_0_8_10", "test-blockchains"]
-test-esplora = ["esplora", "ureq", "electrsd/legacy", "electrsd/esplora_a33e97e1", "test-blockchains"]
+test-esplora = ["electrsd/legacy", "electrsd/esplora_a33e97e1", "test-blockchains"]
 test-md-docs = ["electrum"]
 
 [dev-dependencies]

--- a/macros/src/lib.rs
+++ b/macros/src/lib.rs
@@ -121,3 +121,26 @@ pub fn maybe_await(expr: TokenStream) -> TokenStream {
 
     quoted.into()
 }
+
+/// Awaits if target_arch is "wasm32", uses `tokio::Runtime::block_on()` otherwise
+///
+/// Requires the `tokio` crate as a dependecy with `rt-core` or `rt-threaded` to build on non-wasm32 platforms.
+#[proc_macro]
+pub fn await_or_block(expr: TokenStream) -> TokenStream {
+    let expr: proc_macro2::TokenStream = expr.into();
+    let quoted = quote! {
+        {
+            #[cfg(all(not(target_arch = "wasm32"), not(feature = "async-interface")))]
+            {
+                tokio::runtime::Builder::new_current_thread().enable_all().build().unwrap().block_on(#expr)
+            }
+
+            #[cfg(any(target_arch = "wasm32", feature = "async-interface"))]
+            {
+                #expr.await
+            }
+        }
+    };
+
+    quoted.into()
+}

--- a/src/blockchain/esplora/mod.rs
+++ b/src/blockchain/esplora/mod.rs
@@ -29,38 +29,16 @@ use bitcoin::{BlockHash, Txid};
 use crate::error::Error;
 use crate::FeeRate;
 
-#[cfg(all(
-    feature = "esplora",
-    feature = "reqwest",
-    any(feature = "async-interface", target_arch = "wasm32"),
-))]
+#[cfg(feature = "reqwest")]
 mod reqwest;
 
-#[cfg(all(
-    feature = "esplora",
-    feature = "reqwest",
-    any(feature = "async-interface", target_arch = "wasm32"),
-))]
+#[cfg(feature = "reqwest")]
 pub use self::reqwest::*;
 
-#[cfg(all(
-    feature = "esplora",
-    not(any(
-        feature = "async-interface",
-        feature = "reqwest",
-        target_arch = "wasm32"
-    )),
-))]
+#[cfg(feature = "ureq")]
 mod ureq;
 
-#[cfg(all(
-    feature = "esplora",
-    not(any(
-        feature = "async-interface",
-        feature = "reqwest",
-        target_arch = "wasm32"
-    )),
-))]
+#[cfg(feature = "ureq")]
 pub use self::ureq::*;
 
 fn into_fee_rate(target: usize, estimates: HashMap<String, f64>) -> Result<FeeRate, Error> {
@@ -141,3 +119,11 @@ impl_error!(io::Error, Io, EsploraError);
 impl_error!(std::num::ParseIntError, Parsing, EsploraError);
 impl_error!(consensus::encode::Error, BitcoinEncoding, EsploraError);
 impl_error!(bitcoin::hashes::hex::Error, Hex, EsploraError);
+
+#[cfg(test)]
+#[cfg(feature = "test-esplora")]
+crate::bdk_blockchain_tests! {
+    fn test_instance(test_client: &TestClient) -> EsploraBlockchain {
+        EsploraBlockchain::new(&format!("http://{}",test_client.electrsd.esplora_url.as_ref().unwrap()), 20)
+    }
+}

--- a/src/blockchain/esplora/reqwest.rs
+++ b/src/blockchain/esplora/reqwest.rs
@@ -106,19 +106,19 @@ impl Blockchain for EsploraBlockchain {
     }
 
     fn get_tx(&self, txid: &Txid) -> Result<Option<Transaction>, Error> {
-        Ok(self.url_client._get_tx(txid).await?)
+        Ok(await_or_block!(self.url_client._get_tx(txid))?)
     }
 
     fn broadcast(&self, tx: &Transaction) -> Result<(), Error> {
-        Ok(self.url_client._broadcast(tx).await?)
+        Ok(await_or_block!(self.url_client._broadcast(tx))?)
     }
 
     fn get_height(&self) -> Result<u32, Error> {
-        Ok(self.url_client._get_height().await?)
+        Ok(await_or_block!(self.url_client._get_height())?)
     }
 
     fn estimate_fee(&self, target: usize) -> Result<FeeRate, Error> {
-        let estimates = self.url_client._get_fee_estimates().await?;
+        let estimates = await_or_block!(self.url_client._get_fee_estimates())?;
         super::into_fee_rate(target, estimates)
     }
 }
@@ -287,10 +287,10 @@ impl ElectrumLikeSync for UrlClient {
             for script in chunk {
                 futs.push(self._script_get_history(script));
             }
-            let partial_results: Vec<Vec<ElsGetHistoryRes>> = futs.try_collect().await?;
+            let partial_results: Vec<Vec<ElsGetHistoryRes>> = await_or_block!(futs.try_collect())?;
             results.extend(partial_results);
         }
-        Ok(stream::iter(results).collect().await)
+        Ok(await_or_block!(stream::iter(results).collect()))
     }
 
     fn els_batch_transaction_get<'s, I: IntoIterator<Item = &'s Txid>>(
@@ -303,10 +303,10 @@ impl ElectrumLikeSync for UrlClient {
             for txid in chunk {
                 futs.push(self._get_tx_no_opt(txid));
             }
-            let partial_results: Vec<Transaction> = futs.try_collect().await?;
+            let partial_results: Vec<Transaction> = await_or_block!(futs.try_collect())?;
             results.extend(partial_results);
         }
-        Ok(stream::iter(results).collect().await)
+        Ok(await_or_block!(stream::iter(results).collect()))
     }
 
     fn els_batch_block_header<I: IntoIterator<Item = u32>>(
@@ -319,10 +319,10 @@ impl ElectrumLikeSync for UrlClient {
             for height in chunk {
                 futs.push(self._get_header(height));
             }
-            let partial_results: Vec<BlockHeader> = futs.try_collect().await?;
+            let partial_results: Vec<BlockHeader> = await_or_block!(futs.try_collect())?;
             results.extend(partial_results);
         }
-        Ok(stream::iter(results).collect().await)
+        Ok(await_or_block!(stream::iter(results).collect()))
     }
 }
 

--- a/src/blockchain/esplora/reqwest.rs
+++ b/src/blockchain/esplora/reqwest.rs
@@ -350,11 +350,3 @@ impl ConfigurableBlockchain for EsploraBlockchain {
         Ok(blockchain)
     }
 }
-
-#[cfg(test)]
-#[cfg(feature = "test-esplora")]
-crate::bdk_blockchain_tests! {
-    fn test_instance(test_client: &TestClient) -> EsploraBlockchain {
-        EsploraBlockchain::new(&format!("http://{}",test_client.electrsd.esplora_url.as_ref().unwrap()), None, 20)
-    }
-}

